esp8266-fastled-webserver/common.h

```diff
@@ -0,0 +1,336 @@
+      (ONE_BASED_OUTPUT_CHANNEL == 4) ? PIXELS_ON_DATA_PIN_4 :
+    #endif
+    #if PARALLEL_OUTPUT_CHANNELS >= 3
+      (ONE_BASED_OUTPUT_CHANNEL == 3) ? PIXELS_ON_DATA_PIN_3 :
+    #endif
+    #if PARALLEL_OUTPUT_CHANNELS >= 2
+      (ONE_BASED_OUTPUT_CHANNEL == 2) ? PIXELS_ON_DATA_PIN_2 :
+    #endif
+    #if PARALLEL_OUTPUT_CHANNELS >= 1
+      (ONE_BASED_OUTPUT_CHANNEL == 1) ? PIXELS_ON_DATA_PIN_1 :
+    #endif
+      0;
+  #endif
+}
+
+void writeAndCommitSettings();
+void broadcastInt(String name, uint8_t value);
+void broadcastString(String name, String value);
+
+
+#if defined(ESP32) || defined(ESP8266)
+  // Optional: (LGPL) https://github.com/sinricpro/ESPTrueRandom.git#ed198f459da6d7af65dd13317a4fdc97b23991b4
+  // #include "ESPTrueRandom.h"
+  // Then:
+  //     ESPTrueRandom.useRNG = true;
+  //     int32_t r = ESPTrueRandom.memfill((void*)&r, sizeof(r));
+#else
+  #error "Currently only ESP32 and ESP8266 are supported"
+#endif
+
+
+// Structures
+typedef void (*Pattern)();
+typedef Pattern PatternList[];
+typedef struct {
+  Pattern pattern;
+  String name;
+} PatternAndName;
+
+typedef struct {
+  CRGBPalette16 palette;
+  String name;
+} PaletteAndName;
+
+
+// forward-declarations
+
+// TODO: cleanup to declare minimum necessary variables
+// extern const uint8_t brightnessCount;
+extern const uint8_t brightnessMap[];
+extern uint8_t brightnessIndex;
+extern uint8_t currentPatternIndex;
+extern const uint8_t patternCount;
+extern const PatternAndName patterns[];
+
+extern uint8_t currentPaletteIndex;
+extern uint8_t autoplay;
+extern uint8_t autoplayDuration;
+extern uint8_t showClock;
+extern uint8_t clockBackgroundFade;
+extern CRGB    solidColor;
+extern uint8_t cooling;
+extern uint8_t sparking;
+extern uint8_t speed;
+extern uint8_t twinkleSpeed;
+extern uint8_t twinkleDensity;
+extern uint8_t coolLikeIncandescent;
+extern uint8_t saturationBpm;
+extern uint8_t saturationMin;
+extern uint8_t saturationMax;
+extern uint8_t brightDepthBpm;
+extern uint8_t brightDepthMin;
+extern uint8_t brightDepthMax;
+extern uint8_t brightThetaIncBpm;
+extern uint8_t brightThetaIncMin;
+extern uint8_t brightThetaIncMax;
+extern uint8_t msMultiplierBpm;
+extern uint8_t msMultiplierMin;
+extern uint8_t msMultiplierMax;
+extern uint8_t hueIncBpm;
+extern uint8_t hueIncMin;
+extern uint8_t hueIncMax;
+extern uint8_t sHueBpm;
+extern uint8_t sHueMin;
+extern uint8_t sHueMax;
+
+
+
+extern const TProgmemRGBGradientPalettePtr gGradientPalettes[];
+extern const uint8_t gGradientPaletteCount;
+extern const CRGBPalette16 palettes[];
+extern const uint8_t paletteCount;
+extern const String paletteNames[];
+
+extern CRGBPalette16 gCurrentPalette;
+extern CRGBPalette16 gTargetPalette;
+extern uint8_t gCurrentPaletteNumber;
+
+extern WiFiManager wifiManager;
+extern ESP8266WebServer webServer;
+extern NTPClient timeClient;
+extern String nameString;
+
+extern CRGB leds[NUM_PIXELS];
+
+#if IS_FIBONACCI // actual data in map.h
```

Comment:
There are some external data structures that are only defined when the product supports certain features.  See `map.cpp` for definitions of these.